lib/experimental/class-wp-webfonts.php

```diff
@@ -268,150 +362,439 @@ public function validate_webfont( $webfont ) {
+	private function prep_handles_for_printing( $handles = false ) {
+		if ( false !== $handles ) {
+			$handles = $this->validate_handles( $handles );
+			// Bail out when invalid.
+			if ( null === $handles ) {
+				return array();
+			}
+		}
 
-		// Bail out if there are no styles to enqueue.
-		if ( '' === $styles ) {
-			return;
+		// Use the enqueued queue.
+		if ( empty( $handles ) ) {
+			if ( empty( $this->queue ) ) {
+				trigger_error( 'No web fonts are enqueued for printing' );
+
+				return array();
+			}
+			$handles = $this->queue;
 		}
 
-		// Enqueue the stylesheet.
-		wp_register_style( $this->stylesheet_handle, '' );
-		wp_enqueue_style( $this->stylesheet_handle );
+		return $handles;
+	}
 
-		// Add the styles to the stylesheet.
-		wp_add_inline_style( $this->stylesheet_handle, $styles );
+	/**
+	 * Validates handle(s) to ensure each is a non-empty string.
+	 *
+	 * @since X.X.X
+	 *
+	 * @param mixed $handles Handle or handles to be validated.
+	 * @return string[]|null Array of handles on success. Else null.
+	 */
+	private function validate_handles( $handles ) {
+		// Validate each element is a non-empty string handle.
+		$handles = array_filter(
+			(array) $handles,
+			static function( $handle ) {
+				return is_string( $handle ) && ! empty( $handle );
+			}
+		);
+
+		if ( empty( $handles ) ) {
+			trigger_error( 'Handles must be a non-empty string or array of non-empty strings' );
+			return null;
+		}
+
+		return $handles;
 	}
 
 	/**
-	 * Generate and enqueue editor styles.
+	 * Invokes each provider to process and print its styles.
 	 *
-	 * @since 6.0.0
+	 * @since X.X.X
+	 *
+	 * @see WP_Dependencies::do_item()
+	 *
+	 * @param string    $provider_id The font family to process.
+	 * @param int|false $group       Not used.
+	 * @return bool
 	 */
-	public function generate_and_enqueue_editor_styles() {
-		// Generate the styles.
-		$webfonts = $this->get_webfonts_by_provider( $this->get_all_webfonts() );
-		$styles   = $this->generate_styles( $webfonts );
+	public function do_item( $provider_id, $group = false ) { // phpcs:ignore VariableAnalysis.CodeAnalysis.VariableAnalysis.UnusedVariable
+		// Bail out if the provider is not registered.
+		if ( ! isset( $this->providers[ $provider_id ] ) ) {
+			return false;
+		}
 
-		// Bail out if there are no styles to enqueue.
-		if ( '' === $styles ) {
-			return;
+		$font_handles = $this->get_enqueued_fonts_for_provider( $provider_id );
+		if ( empty( $font_handles ) ) {
+			return false;
+		}
+
+		$properties_by_font = $this->get_font_properties_for_provider( $font_handles );
+		if ( empty( $properties_by_font ) ) {
+			return false;
+		}
+
+		// Invoke provider to print its styles.
+		if ( isset( $this->provider_instances[ $provider_id ] ) ) {
+			$provider = $this->provider_instances[ $provider_id ];
+		} else {
+			$provider = new $this->providers[ $provider_id ]['class']();
+			// Store the instance.
+			$this->provider_instances[ $provider_id ] = $provider;
 		}
```

Comment:
I believe we can simplify this a bit...
```suggestion
		if ( ! isset( $this->provider_instances[ $provider_id ] ) ) {
			$this->provider_instances[ $provider_id ] = new $this->providers[ $provider_id ]['class']();
		}
		$provider = $this->provider_instances[ $provider_id ];
```

Reply:
Simplified and encapsulated in commit https://github.com/WordPress/gutenberg/pull/43492/commits/c874aac178ca9b28708cce85e6ec48b974cd05fc.

Great suggestion @aristath!